src/dhtproto/client/legacy/internal/helper/Handshake.d

```diff
@@ -0,0 +1,241 @@
+            been reached
+
+        ***********************************************************************/
+
+        private void allNodesConnected ()
+        {
+            this.outer.log.info("All DHT nodes connected!");
+            this.all_connected_event.trigger();
+        }
+
+        /***********************************************************************
+
+            Runs the task
+
+        ***********************************************************************/
+
+        protected override void run ()
+        {
+            this.all_connected_event.wait();
+        }
+    }
+
+
+    /***************************************************************************
+
+        Task that suspends until at least one node is connected
+
+    ***************************************************************************/
+
+    private class OneDhtNodeConnected : Task
+    {
+        import swarm.Const : NodeItem;
+
+        /// Used to manage suspend/resume of this task
+        private TaskEvent node_connected_event;
+
+        /***********************************************************************
+
+            Delegate to be called by the handshake each time an individual
+            node connects
+
+            Params:
+                item = data on the address and port of the connected node
+
+        ***********************************************************************/
+
+        private void oneNodeConnected (NodeItem item)
+        {
+            this.outer.log.trace("Connected to DHT node {}:{}",
+                                 item.Address, item.Port);
+            this.node_connected_event.trigger();
+        }
+
+        /***********************************************************************
+
+            Runs the task
+
+        ***********************************************************************/
+
+        protected override void run ()
+        {
+            this.node_connected_event.wait();
+        }
+    }
+
+
+    /// Logger for the handshake
+    private Logger log;
+
+
+    /// Instance of task that suspends until all nodes are connected
+    private AllDhtNodesConnected all_nodes_connected;
+
+
+    /// Instance of task that suspends until at least one node is connected
+    private OneDhtNodeConnected one_node_connected;
+
+
+    /// Instance of class that launches the actual handshake, retrying when
+    /// it fails until all nodes are connected
+    private RetryHandshake retry_handshake;
+
+
+    /***************************************************************************
+
+        Constructor
+
+        Params:
+            dht = DHT client instance via which to perform the handshake
+            retry_delay_seconds = time to wait (in seconds) before retrying
+                failed handshake attempts
+
+    ***************************************************************************/
+
+    public this (DhtClient dht, size_t retry_delay_seconds)
+    in
+    {
+        assert(dht !is null);
```

Comment:
Could you check here that the scheduler has really been initialised? (i.e. that the epoll instance to be used is valid.)

Reply:
Hmmm, is there a mechanism for this in the ocean version dhtproto 13.x.x needs to support?

More recent ocean versions define an `isSchedulerUsed` function that can be used to check that, but that will only be available from dhtproto 14.x.x onwards.

Note that `theScheduler` has an out-contract that verifies that the underlying `Scheduler` instance has been initialized (i.e. it can't return `null`, and therefore in turn `initScheduler` must have been called).

Reply:
Oh that's fine then :+1:.

Reply:
Cool!  Let me know once it's merged into 14.x.x: it may be worth tweaking it to use `IScheduler`, and we _can_ use `isSchedulerUsed()` as an extra check if wanted.